### ISSUE TITLE
Pod spec validation failing with Xcode 12

### ISSUE THREAD

**Issue by mokagio** [Automattic/Automattic-Tracks-iOS#149]
I tried to setup Xcode 12 in order to use the `XCTSkip` API and started getting a failure from the `pod spec lint` task on CI. For example, [this](https://app.circleci.com/pipelines/github/Automattic/Automattic-Tracks-iOS/372/workflows/bfdd2398-2a18-43a5-b8d1-bfcaf3f80ebc/jobs/903), [this](https://app.circleci.com/pipelines/github/Automattic/Automattic-Tracks-iOS/373/workflows/5b11b710-7f22-4eb5-8244-ef746dfc1ac1/jobs/906), and [this](https://app.circleci.com/pipelines/github/Automattic/Automattic-Tracks-iOS/375/workflows/171518d5-2b4d-457f-939f-6a0e76b2cc77/jobs/911).

The final log from the `pod spec lint` command, filtered from all the "The iOS Simulator deployment target '`IPHONEOS_DEPLOYMENT_TARGET`' is set to 8.0, but the range of supported deployment target versions is 9.0 to 14.0.99" is:

```
 -> Automattic-Tracks-iOS (0.5.1)
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code. You can use `--verbose` for more information.
    - NOTE  | xcodebuild:  note: Using new build system
    - NOTE  | xcodebuild:  note: Building targets in parallel
    - NOTE  | xcodebuild:  note: Using codesigning identity override: -
    - NOTE  | [iOS] xcodebuild:  note: Planning build
    - NOTE  | [iOS] xcodebuild:  note: Constructing build description
    - NOTE  | [iOS] xcodebuild:  clang: error: linker command failed with exit code 1 (use -v to see invocation)
    - NOTE  | [iOS] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
```

Unfortunately, it doesn't report the output of `xcodebuild`.

I tried it locally, and tracked it down to the iOS platform target. Running the lint command with `--platforms=macos` works, but `--platforms=ios` fails.

We can then try to replicate the failure by using the `--no-clean` option, which leaves the temp folder where it put the code-generated `App.xcworkspace` it used for the validation for further inspection:

```
bundle exec pod spec --platforms=ios --no-clean
```

That workspace builds for me in Xcode, and also with `xcodebuild` if I point it to an iOS Simulator to avoid code signing.

I _think_ the issue has to do with the Sodium library. The following command from the CocoaPods lint workspace gives the following error:

```
xcodebuild build \
  -workspace App.xcworkspace \
  -scheme App \
  -configuration Release \
  -destination 'platform=iOS Simulator,name=iPhone 11' \
 | xcbeautify
```

```
❌ ld: in /path/to/CocoaPods-Lint/workspace/Pods/Sodium/Sodium/libsodium/libsodium-ios.a(libsodium_la-aead_xchacha20poly1305.o), 
building for iOS Simulator, but linking in object file built for iOS,
for architecture arm64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I'm not sure where to go from here, ~~but the first thing I'd try to do is updating the library~~ Sodium is already at the latest version, 0.8.0.

Any thoughts?

**Comment by mokagio** [comment]
From what I can make from the error, the issue is that during linking, `ld`, the `.a` library was build for iOS arm64 while the build process was targeted at the iOS Simulator.

**Comment by mokagio** [comment]
[This](https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios) might be useful.

**Comment by mokagio** [comment]
[This](https://github.com/CocoaPods/CocoaPods/issues/10059#issuecomment-693346078) has an informative discussion, too.

**Comment by mokagio** [comment]
I'm biased right now because this issue is blocking me, but from what I can make of it, the problem is out of our hands.

Since Xcode 12 is meant for Big Sur, which paves the way for Apple Silicon, the way Xcode builds and selects its architectures has changed. It's now up to library maintainers to update their `.podspec`s to correctly exclude `arm64`. I'm not sure how to do it for a static library like the `libSodium.a` that's giving us problems, but in the issue before this comment someone posted a Gist to [strip invalid architectures from dynamic libraries](https://gist.github.com/artemch/9cd6a467ad1177a73d08e55f8ed789e4) which could be a starting point.

That's all to say, I think we have the tools to work around this issue as consumers of pods in an app and as authors of pods for the portion of settings that we can control, but I haven't been able to find a way to get the lint command to pick up the configuration changes.

Maybe disabling the validation for now could be a viable option? What do you think @jleandroperez @jkmassel ?

Although... are we going to be able to publish a new version if the `.podspec` doesn't validate? 🤔

**Comment by jleandroperez** [comment]
@mokagio Q for you sir! How did you manage to run this? Did you set a breakpoint somewhere and grab Cocoapods internal linting project?

```
xcodebuild build \
  -workspace App.xcworkspace (...)
```

I've been tinkering a bit, buuut at this stage, honestly, I'd very much need to mess with the project that yields this error.
Only way in which ARM should be picked for iOS was either... running on the device (not the case!), or ARM Mac + Simulator, which isn't either the case.

If it's an ARM64 static lib being picked up, it's super odd, since all of the targets have Catalyst (iOS on Mac!) disabled.

As per disabling validation, prooooobably a bad idea (if it's even possible!).

**Comment by mokagio** [comment]
Thanks for looking into this @jleandroperez 🙏 

### To reproduce

1. I run `bundle exec pod spec --platforms=ios --no-clean` from this repo, and got a lint failure

![image](https://user-images.githubusercontent.com/1218433/95158943-8c256480-07e8-11eb-8648-47de27730ae4.png)

2. Went into the folder that CocoaPods generated for its test, see "Pods workspace available at..." in the output above, which was left available because of the `--no-clean` option

3. I run a clean build with `xcodebuild` in `Release` mode

```
xcodebuild build \
  -workspace App.xcworkspace \
  -scheme App \
  -configuration Release \
  -sdk iphonesimulator \
  -destination 'platform=iOS Simulator,name=iPhone 11' \
 | xcbeautify
```

and got this error

![image](https://user-images.githubusercontent.com/1218433/95159119-f9d19080-07e8-11eb-86d4-5cfa9c1bceb9.png)

### Regarding disabling the lint check

Yeah, it's would really be a "stick your head in the sand" kind of thing to do. Also, even if we could skip it for PRs, I doubt we could publish a new version if the spec doesn't pass the linting, so there's not much to do... other than finding a fix.

**Comment by mokagio** [comment]
I've been able to fix the validation for version 0.8.0 of the Sodium pod in [my fork](https://github.com/mokagio/swift-sodium/commit/57b0caf54439cef986ff1adbc58b82f2f39fe353) by following the suggestions in the [StackOverflow question linked above](https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios).

```diff
 s.pod_target_xcconfig = {
    'SWIFT_INCLUDE_PATHS' => '$(PODS_TARGET_SRCROOT)/Sodium/libsodium',
+   'EXCLUDED_ARCHS[sdk=iphonesimulator*]' => 'arm64'
+}
+
+s.user_target_xcconfig = {
+   'EXCLUDED_ARCHS[sdk=iphonesimulator*]' => 'arm64'
 }
 
 s.requires_arc = true
```

With the Xcode 12.0 toolchain, if you checkout that commit and run `pod spec lint --platforms=ios,macos ./Sodium.podspec` (no `bundle exec` because Sodium doesn't use Bundler) it will succeed. Conversely, if you checkout the [`0.8.0`](https://github.com/Automattic/Automattic-Tracks-iOS/blob/5f96537a62a8b7d3eb6635cb4dd8489dd39fc62a/Automattic-Tracks-iOS.podspec#L35) tag (the one currently used by Tracks) and run the same command, it will fail.

There are two problems:

- There is no way I can find to use private Pods as `.podspec` dependencies
- The Sodium maintainers are working on a new version that use an `.xcframework` and I haven't been able to make that version validate yet

**Comment by jleandroperez** [comment]
@mokagio apologies I couldn't get back to this one today!! 

One thing we **could** do is set a custom `Sodium` "source repository". But that has to be done at the "Top Project" level (that is, WPiOS, WCiOS and Simplenote).

Another maybe cooler approach is to... drop Sodium perhaps. But I suspect doing so would be dead tough!!

We could also fork Sodium, fix all the things there, and publish our patched Podspec, until the official maintainers ship a fixed version. WDYT?

(I did something like that a while ago for the Google Authenticator dependency!)

**Comment by mokagio** [comment]
> We could also fork Sodium, fix all the things there, and publish our patched Podspec, until the official maintainers ship a fixed version. WDYT?

That's what I had in mind as a quick fix yesterday and I implemented it today in #148.
